### PR TITLE
PLAT-9256: make sure thumbnail actions are cached for KES rules

### PR DESCRIPTION
**Pull request by noam-arad**

### CODE REVIEW

File: alpha/lib/model/accessControl.php [hunk 1]
```diff
@@ -156,7 +156,20 @@ public function hasRules($contextType = null)
...
-	
+
+	public function hasKESRule()
+	{
+		$rules = $this->getRulesArray();
+
+		foreach($rules as $rule)
+		{
+			/* @var $rule kRule */
+			if (json_decode($rule->getRuleData())->type == 'ServeFromKESRule')
```
Comment (kalturaguy) [inline alpha/lib/model/accessControl.php]: i think it's better to check if one of the action is KalturaAccessControlServeRemoteEdgeServerAction

+ return false at the end of the method

Comment (kalturaguy) [reply alpha/lib/model/accessControl.php]: also for performance issue, let's save the result in the custom_data when you save the ACP, and just check the custom_data in the action

File: alpha/lib/model/accessControl.php [hunk 2]
```diff
@@ -309,4 +330,37 @@ public function getCacheInvalidationKeys()
...
+		foreach ($rules as $rule)
+		{
+			/* @var $rule kRule */
+			$actions = $rule->getActions();
+			foreach ($actions as $action)
+			{
+				/* @var kRuleAction $action */
+				if ($action->getType() == RuleActionType::SERVE_FROM_REMOTE_SERVER)
+				{
+					$hasKESRule = true;
```
Comment (yossipapi) [inline alpha/lib/model/accessControl.php]: You can probably break from the loop after hasKESRule was assigned with true.

File: alpha/lib/model/accessControl.php [hunk 3]
```diff
@@ -309,4 +330,37 @@ public function getCacheInvalidationKeys()
...
+
+	public function setHasKES($v)
+	{
+		$this->putInCustomData(self::CUSTOM_DATA_HAS_KES_RULE, $v);
+	}
+
+	/**
+	 * @return bool
+	 */
+	protected function calculateHasKESRule(): bool
```
Comment (yossipapi) [inline alpha/lib/model/accessControl.php]: Remove the ": bool" from the function signature

File: alpha/apps/kaltura/modules/extwidget/actions/thumbnailAction.class.php [hunk 4]
```diff
@@ -453,8 +453,8 @@ public function execute()
 		}
 		
 		$nocache = false;
-		if ($secureEntryHelper->shouldDisableCache() || kApiCache::hasExtraFields() ||
-			(!$secureEntryHelper->isKsWidget() && $secureEntryHelper->hasRules(ContextType::THUMBNAIL)))
+		if (!$secureEntryHelper->hasKESRule() && ($secureEntryHelper->shouldDisableCache() || kApiCache::hasExtraFields() ||
```
Comment (yossipapi) [inline alpha/apps/kaltura/modules/extwidget/actions/thumbnailAction.class.php]: I would probably change all the places where you named the functions or variables from the word KES to remoteServer (i.e hasKESRule hasServeRemoteServerRule). KES is not a terminology the Kaltura server should know about.

File: alpha/lib/model/accessControl.php [hunk 5]
```diff
@@ -309,4 +330,38 @@ public function getCacheInvalidationKeys()
 	{
 		return array("accessControl:id=".strtolower($this->getId()));
 	}
+
+	public function getHasRemoteServer()
```
Comment (yossipapi) [inline alpha/lib/model/accessControl.php]: The function name is not that suitable, should be something like getHasServeRemoteServerRule same applies to other method signature as well.

**Pull request merged**